pkg/internalregistry/controllers/service_account_controller.go

```diff
@@ -134,16 +139,56 @@ func (c *serviceAccountController) sync(ctx context.Context, key string) error {
 	// name of managed secret
 	secretName, err := c.managedImagePullSecretName(ctx, serviceAccount)
 
-	// ensure secret ref annotation is set
-	if serviceAccount.Annotations[InternalRegistryImagePullSecretRefKey] != secretName {
-		patch, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
-		if err != nil {
-			return err
+	// extract the apply configuration patch from the cache
+	patch, err := applycorev1.ExtractServiceAccount(serviceAccount, serviceAccountControllerFieldManager)
+	if err != nil {
+		return err
+	}
+
+	// we always want the annotation set
+	patch.WithAnnotations(map[string]string{InternalRegistryImagePullSecretRefKey: secretName})
```

Comment:
I think we want to "assume an open world" and build the desired apply configuration from scratch rather than by modifying the existing apply configuration. As an added benefit, that should remove the need for some of the conditional branches in the control loop, too.

Reply:
I purposely choose to extracted the patch. Comparing the extracted `ImagePullSecrets` to the cached `ImagePullSecrets` allows me to detect the case where we do not own the field, yet there is no need to update the `ImagePullSecrets`.

Reply:
Refactored.